code/modules/flufftext/Hallucination.dm

```diff
@@ -242,34 +251,64 @@ GLOBAL_LIST_INIT(hallucination_list, list(
 		target.Paralyze(100)
 		target.visible_message("<span class='danger'>[target] flails around wildly.</span>","<span class='userdanger'>[name] pounces on you!</span>")
 
+// The numbers of seconds it takes to get to each stage of the xeno attack choreography
+#define XENO_ATTACK_STAGE_LEAP_AT_TARGET 1
+#define XENO_ATTACK_STAGE_LEAP_AT_PUMP 2
+#define XENO_ATTACK_STAGE_CLIMB 3
+#define XENO_ATTACK_STAGE_FINISH 6
+
+/// Xeno crawls from nearby vent,jumps at you, and goes back in
 /datum/hallucination/xeno_attack
-	//Xeno crawls from nearby vent,jumps at you, and goes back in
-	var/obj/machinery/atmospherics/components/unary/vent_pump/pump = null
+	var/turf/pump_location = null
 	var/obj/effect/hallucination/simple/xeno/xeno = null
+	var/time_processing = 0
+	var/stage = XENO_ATTACK_STAGE_LEAP_AT_TARGET
 
 /datum/hallucination/xeno_attack/New(mob/living/carbon/C, forced = TRUE)
-	set waitfor = FALSE
 	..()
 	for(var/obj/machinery/atmospherics/components/unary/vent_pump/U in orange(7,target))
 		if(!U.welded)
-			pump = U
+			pump_location = get_turf(U)
 			break
-	if(pump)
-		feedback_details += "Vent Coords: [pump.x],[pump.y],[pump.z]"
-		xeno = new(pump.loc,target)
-		sleep(10)
-		xeno.update_icon("alienh_leap",'icons/mob/alienleap.dmi',-32,-32)
-		xeno.throw_at(target,7,1, xeno, FALSE, TRUE)
-		sleep(10)
-		xeno.update_icon("alienh_leap",'icons/mob/alienleap.dmi',-32,-32)
-		xeno.throw_at(pump,7,1, xeno, FALSE, TRUE)
-		sleep(10)
-		var/xeno_name = xeno.name
-		to_chat(target, "<span class='notice'>[xeno_name] begins climbing into the ventilation system...</span>")
-		sleep(30)
-		qdel(xeno)
-		to_chat(target, "<span class='notice'>[xeno_name] scrambles into the ventilation ducts!</span>")
-	qdel(src)
+
+	if(pump_location)
+		feedback_details += "Vent Coords: [pump_location.x],[pump_location.y],[pump_location.z]"
+		xeno = new(pump_location, target)
+		START_PROCESSING(SSfastprocess, src)
+	else
+		qdel(src)
+
+/datum/hallucination/xeno_attack/process(delta_time)
+	time_processing += delta_time
+
+	if (time_processing >= stage)
+		switch (time_processing)
+			if (XENO_ATTACK_STAGE_FINISH to INFINITY)
+				to_chat(target, "<span class='notice'>[xeno.name] scrambles into the ventilation ducts!</span>")
+				qdel(src)
+			if (XENO_ATTACK_STAGE_CLIMB to XENO_ATTACK_STAGE_FINISH)
+				to_chat(target, "<span class='notice'>[xeno.name] begins climbing into the ventilation system...</span>")
+				stage = XENO_ATTACK_STAGE_FINISH
+			if (XENO_ATTACK_STAGE_LEAP_AT_PUMP to XENO_ATTACK_STAGE_CLIMB)
+				xeno.update_icon("alienh_leap",'icons/mob/alienleap.dmi', -32, -32)
+				xeno.throw_at(pump_location, 7, 1, xeno, FALSE, TRUE)
+				stage = XENO_ATTACK_STAGE_CLIMB
+			if (XENO_ATTACK_STAGE_LEAP_AT_TARGET to XENO_ATTACK_STAGE_LEAP_AT_PUMP)
+				xeno.update_icon("alienh_leap",'icons/mob/alienleap.dmi', -32, -32)
+				xeno.throw_at(target, 7, 1, xeno, FALSE, TRUE)
+				stage = XENO_ATTACK_STAGE_LEAP_AT_PUMP
```

Comment:
couple things, first of all why are you operating on both stages and a switch statement? Shouldn't you switch on stage, rather then on time_processing?

I'm not really too big a fan of how the stage timing works, a comment or something would be nice.

Reply:
It's due to the assumption that delta time can be longer than the time it takes to go from stage 1 -> stage 2, i.e. if stage 1 happens at 1 second, stage 2 happens at 2 seconds, stage 3 happens at 3 seconds, and you get a delta time of 3 seconds, I want stage 1 and 2 to be skipped entirely.

In practice, this does not currently happen, but one of the purposes of having delta time in the first place is this kind of forwards compatibility.

`stage` is kept in order to make sure we don't run the same stage code multiple times if the delta time is less than the time it takes to go from stage A to stage B.

Reply:
👍